### PR TITLE
feat: adding support for data uri sanitisation

### PR DESCRIPTION
**Pull request by sahil-seth**
## Changes

This PR aims to address this feature request https://github.com/renovatebot/renovate/discussions/31719 where we can now sanitise possible data uris coming in the logs.

## Context

- #31719

## Documentation (please check one with an [x])

- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [x] Newly added/modified unit tests, or

### CODE REVIEW

File: lib/logger/utils.ts
Comment (rarkins) [inline lib/logger/utils.ts:332]: Does all data really need replacing?

Comment (sahil-seth) [reply lib/logger/utils.ts:332]: IMO, yes cause the part after data may contain sensitive information such as the issuer of the uri along with the key itself. This is what we try to follow in our firm as well.

**Comment by sahil-seth** [comment]
@rarkins should we go ahead with this change? 
The workflows seem to be waiting an approval from a maintainer

**Review by rarkins — DISMISSED** [review]
These Regex seem too aggressive and will match some unintentional/irrelevant data messages too

**Comment by sahil-seth** [comment]
@rarkins Good point, but do you have any examples that you can think of, which might match to the new regex but should not ideally?
Asking cause it might help me optimise the regex quicker.

**Comment by rarkins** [comment]
No, because there are infinite examples possible, so you should provide some examples of where data uri's need redacting and we make the Regex tighter. For example should it occur only at the start of a field?

**Comment by sahil-seth** [comment]
Here are some data URI examples which I aimed to redact with this change:

`data:application/pkcs8;kid=example-key-123;base64,MIIBIjANBgkqhkiG9w0BAp1s1F3lQfZ9c/GbE=` 
`data:application/jwt;kid=jwt-key-789;base64,eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9`
`data:application/x-pem-file;kid=pem-cert-202;base64,LS0tLS1CZWdpbiBDRVJUSUZJQ0FURS0tLS0t`
`data:audio/wav;kid=audio-key-707;base64,UklGRjIAAABXQVZFZm10IBAAAAABAAEAESsAACJWAAACABAAZGF0YYAAAA`

Also, upon a quick check with an AI assistant, the regex doesn't seem to match unnecessary strings apart from URLs or data URI like fields.

**Comment by sahil-seth** [comment]
> For example should it occur only at the start of a field?

Yes, it can. It might be safer to redact the whole string after `data` cause it takes away the knowledge of what kind of key it is, whether it is `jwt` or `pkcs8` or `pem-file` along with the key itself.

WDYT?

**Comment by rarkins** [comment]
Please pricier a full log example from today (you can redact parts of it yourself). Eg where does this appear today?

**Comment by sahil-seth** [comment]
Screenshot 2024-10-01 at 12 28 30 PM

Here is an example

**Comment by rarkins** [comment]
Great, so can we start with the case where we match against `^data:` and it's the whole value of the log message? (as in your examples)

**Comment by sahil-seth** [comment]
@rarkins I'm not sure If I got you this time, could you please elaborate on the suggestion?

**Comment by sahil-seth** [comment]
@rarkins just following up on above message, 
do you mean we update the regex to include the strings starting with `data:` ?

**Comment by rarkins** [comment]
Yes, the regex should be strict to match only where the log content starts and ends with a data URI

**Comment by viceice** [comment]
we can even more tighten the regex to also require the content-type after `data:`

`^data:[0-9a-z-]+/[0-9a-z-]+;.+`

**Comment by sahil-seth** [comment]
@rarkins @viceice 
Just to clarify and make sure we are on the same page,

Are we talking about `dataUriCredRe` OR `urlRe` ?

File: lib/logger/utils.ts
Comment (rarkins) [inline lib/logger/utils.ts:325]: Revert this change, instead add extra "if" logic later to specifically handle dataUri's

**Comment by rarkins** [comment]
It's better if you revert your changes to `urlRe`

**Comment by sahil-seth** [comment]
@rarkins I've reverted the change to `urlRe` and added a "if" case for data URIs, LMK if this looks fine?

File: lib/logger/utils.ts
Comment (sahil-seth) [reply lib/logger/utils.ts]: Done

File: lib/logger/utils.ts
Comment (viceice) [inline lib/logger/utils.ts:327]:
```suggestion
const dataUriCredRe = /^(data:[0-9a-z-]+\/[0-9a-z-]+;).+/i;
```

File: lib/logger/utils.ts
Comment (viceice) [inline lib/logger/utils.ts:340]:
```suggestion
  if (sanitizedText.startsWith('data:')) {
    sanitizedText = sanitizedText.replace(dataUriCredRe, (dataUri) => {
      return 'data:**redacted**';
    });
  }
  return sanitizedText.replace(dataUriCredRe, '$1**redacted**');
```
simplify

File: lib/logger/utils.ts
Comment (viceice) [inline lib/logger/utils.ts:330]:
```suggestion
  return text.replace(urlRe, (url) => {
```

File: lib/logger/utils.ts
Comment (viceice) [inline lib/logger/utils.ts:340]:
```suggestion
   }).replace(dataUriCredRe, '$1**redacted**');
```
needs prettifying

**Comment by viceice** [comment]
please run prettier to fix linter

**Comment by sahil-seth** [comment]
@viceice done!

**Comment by sahil-seth** [comment]
@rarkins @viceice Does this PR looks good to merge now?
once the pipelines are triggered...

**Review by viceice — APPROVED** [review]

**Comment by sahil-seth** [comment]
@rarkins awaiting your review/approval here

**Comment by sahil-seth** [comment]
@rarkins @viceice looks like the pipelines need to be run once again?

**Comment by viceice** [comment]
> @rarkins @viceice thanks for enabling auto-merge but looks like the pipelines need to be run once again?

stop merging main, we're using GitHub merge queue. if you push something then actions need new approval because you're a first time contributer

**Pull request merged**
